packages/workbox-background-sync/src/lib/QueueDb.ts

```diff
@@ -0,0 +1,149 @@
+/*
+  Copyright 2018 Google LLC
+
+  Use of this source code is governed by an MIT-style
+  license that can be found in the LICENSE file or at
+  https://opensource.org/licenses/MIT.
+*/
+
+import {openDB, DBSchema, IDBPDatabase} from 'idb';
+import {RequestData} from './StorableRequest.js';
+
+interface QueueDBSchema extends DBSchema {
+  requests: {
+    key: number;
+    value: UnidentifiedQueueStoreEntry;
+    indexes: {queueName: string};
+  };
+}
+
+const DB_VERSION = 3;
+const DB_NAME = 'workbox-background-sync';
+const OBJECT_STORE_NAME = 'requests';
+const INDEXED_PROP = 'queueName';
+
+export interface UnidentifiedQueueStoreEntry {
+  requestData: RequestData;
+  timestamp: number;
+  id?: number;
+  queueName?: string;
+  metadata?: object;
+}
+
+export interface QueueStoreEntry extends UnidentifiedQueueStoreEntry {
+  id: number;
+}
+
+/**
+ * A class to interact directly with IndexedDB to store a retrieve queue requests.
+ *
+ * @private
+ */
+
+export class QueueDb {
+  private _db: IDBPDatabase<QueueDBSchema> | null = null;
+
+  /**
+   * Opens a connection to IDBDatabase.
+   *
+   * @private
+   */
+  private async open() {
+    if (!this._db) {
+      this._db = await openDB(DB_NAME, DB_VERSION, {
+        upgrade: this._upgradeDb,
+      });
+    }
+  }
+
+  /**
+   *
+   * @param {UnidentifiedQueueStoreEntry} entry
+   */
+  async addEntry(entry: UnidentifiedQueueStoreEntry) {
+    await this.open();
+    await this._db!.add(OBJECT_STORE_NAME, entry);
+  }
+
+  /**
+   * Returns the first entry in the ObjectStore
+   *
+   * @return {QueueStoreEntry}
+   */
+  async getFirstEntry(): Promise<QueueStoreEntry | any> {
+    await this.open();
+    const cursor = await this._db
+      ?.transaction(OBJECT_STORE_NAME)
+      .store.openCursor();
+
+    if (cursor) {
+      return cursor.value as QueueStoreEntry;
+    }
+  }
+
+  /**
+   *
+   * @param query
+   * @return {Promise<QueueStoreEntry[]>}
+   */
+  async getAllEntriesFromIndex(
+    query?: IDBKeyRange,
+  ): Promise<QueueStoreEntry[]> {
+    await this.open();
+    const results =
+      await this._db?.getAllFromIndex(OBJECT_STORE_NAME, INDEXED_PROP, query) as QueueStoreEntry[];
+    return results ? results : new Array<QueueStoreEntry>();
+  }
+
+  /**
+   *
+   * @param {number} id the id of the entry to be deleted
+   */
+  async deleteEntry(id: number) {
+    await this.open();
+    await this._db?.delete(OBJECT_STORE_NAME, id);
+  }
+
+  /**
+   * Returns either the first or the last entries, depending on direction. Uses the index.
+   *
+   * @param {IDBCursorDirection} direction
+   * @param {IDBKeyRange} query
+   * @return {Promise<QueueStoreEntry | any>}
+   */
+  async getEndEntryFromIndex(
+    {direction}: {direction?: IDBCursorDirection},
```

Comment:
It seems a little awkward to have one parameter as a destructured object, and the second parameter as just a normal object.

Since I think this is not exposed publicly, do you want to just change the method signature so that it's either a single parameter that gets destructured, or alternatively two parameters without any destructuring? So:

```
async getEndEntryFromIndex({direction, query}: {
  direction?: IDBCursorDirection,
  query?: IDBKeyRange,
})
```

or

```
async getEndEntryFromIndex(direction?: IDBCursorDirection, query?: IDBKeyRange)
```

Also, if you know that one or both of the parameter values will always be set, you don't need to add a `?` after them. It's a TypeScript best practice to be as strict as possible and only loosen up restrictions (via `?`) if there are times when you really don't know if the value will be `undefined`.

Reply:
Yes, copy pasted from QueueStore, restructured all this for better readability + clearer names.